### PR TITLE
[MRG+1] Expose logging function in joblib.logger

### PR DESCRIPTION
**Pull request by lesteve**
This way it can be overriden globally via monkey patching.

The use case I am facing would be to replace printing to stdout by writing to a log file.

### CODE REVIEW

File: joblib/__init__.py [hunk 1]
```diff
@@ -58,7 +58,10 @@
    inputs and  outputs: Python functions. Joblib can save their
    computation to disk and rerun it only if necessary::
 
+
       >>> from joblib import Memory
+      >>> import shutil ; import os
+      >>> if os.path.exists('/tmp/joblib'): shutil.rmtree('/tmp/joblib/')
```
Comment (GaelVaroquaux) [inline joblib/__init__.py]: I don't really like that it makes the docstring harder to read. Maybe it could be moved elsewhere.

**Comment by GaelVaroquaux** [comment]
Cheeky mode: why don't we use Python's logging framework inside the log function?

File: joblib/__init__.py (hunk 1 above)
Comment (lesteve) [inline joblib/__init__.py]: Sorry I did not mean to change that. I had a problem with the doctests that were not passing so I was trying to debug. I'll look at my diff in more details.

**Comment by lesteve** [comment]
> Cheeky mode: why don't we use Python's logging framework inside the log function?

I guess you can monkey-patch `joblib.logger.log` and use stdlib logging if you feel like it :).

**Comment by GaelVaroquaux** [comment]
> I guess you can monkey-patch joblib.logger.log and use stdlib logging if you
> feel like it :).

But the logging module makes it very easy to write to a file.

I mean, my question wasn't that cheeky: many Python programmer would
expect the logging module.

**Comment by lesteve** [comment]
In my use case, the stdout is lost into the ether and I just want to override the `print` function globally in order to have the output from Parallel into some file logs.

This PR is just the simplest and non-controversial way I could think of to achieve that.

I feel discussions on whether to use stdlib logging are more longer term discussions. I guess if we do that we would need to do it for scikit-learn as well, which as you may know is not the most uncontroversial issue in the world (how to support decreasing verbose in nested calls jumps to mind for example).

**Comment by lesteve** [comment]
Hmmm I just realised that to be monkey-patching friendly, you need to import the `logger` module and call `logger.log` rather than the `log` function.

**Comment by lesteve** [comment]
I was thinking about adding a test monkey-patching `joblib.logger.log`. Should be easily doable with pytest.

**Comment by aabadie** [comment]
> I was thinking about adding a test monkey-patching joblib.logger.log. Should be easily doable with pytest.

Sounds good. Looking forward for this new test :)

**Comment by karandesai-96** [comment]
Monkey patching tests are easier than ever with pytest :wink: ( Following this thread )

**Comment by lesteve** [comment]
Tests added.

**Comment by aabadie** [comment]
This PR needs rebase @lesteve. I think we can move forward fast on this one.

**Comment by lesteve** [comment]
I think this one was not deemed worth the pain by @ogrisel at one point ...

**Comment by pierreglaser** [comment]
I think that would be great to revive this PR:

- First, this would allow us to close once and for all #692 
- Moreover, the two main `features` of joblib when compared to lower-level executor-based libraries (`loky`, `concurrent.futures`, `dask`...), e.g smart batching as well as efficient file transfer, would be way more gracefully introspected using isolated logging instances that could be toggled when desired.

**Comment by lesteve** [comment]
Unlikely to work on this anytime soon, closing.

**Pull request closed without merging**